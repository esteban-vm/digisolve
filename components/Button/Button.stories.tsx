import { type Meta, type StoryObj, within, userEvent } from '@/stories'
import Button from './Button'

export default {
  args: {
    text: 'button example',
  },
  component: Button,
  title: 'Components/Button',
  parameters: {
    layout: 'centered',
    viewport: {
      disable: true,
    },
    controls: {
      exclude: /^(is(Full|Submit)|link|theme|className)$/,
    },
  },
  play({ canvasElement }) {
    const canvas = within(canvasElement)
    const button = canvas.getByRole('button')
    userEvent.click(button)
  },
} as Meta<typeof Button>

export const BasicExample: StoryObj<typeof Button> = {
  args: { isFull: false },
}

export const FullExample: StoryObj<typeof Button> = {
  args: { isFull: true },
}

export const SubmitExample: StoryObj<typeof Button> = {
  args: { isSubmit: true, isFull: true },
  argTypes: { isSubmit: { control: false } },
}
