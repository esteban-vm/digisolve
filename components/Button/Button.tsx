import type { Component, PropsWithLink } from '@/types'
import { animateScroll as scroll, scroller } from 'react-scroll'
import { styled, css } from '@/styles'

type ButtonProps = Omit<PropsWithLink<{ text: string; isSubmit?: boolean; onClick?: () => void }>, 'id'>

const ButtonComponent: Component<ButtonProps> = ({ text, isSubmit, onClick, link, className }) => {
  const clickHandler = () => {
    if (onClick) {
      onClick()
    } else {
      if (link) scroller.scrollTo(link, { smooth: 'easeInQuint' })
      else scroll.scrollToTop({ smooth: 'easeOutQuint' })
    }
  }

  return (
    <button type={isSubmit ? 'submit' : 'button'} onClick={isSubmit ? undefined : clickHandler} className={className}>
      {text}
    </button>
  )
}

type StyledButtonProps = { isFull?: boolean }

const Button = styled(ButtonComponent)<StyledButtonProps>`
  cursor: pointer;
  display: inline-block;
  text-decoration: none;
  text-transform: capitalize;
  padding: 10px 30px;
  border-radius: 200px;
  font-weight: 500;
  outline: none;
  transition: background-color 500ms;
  border: 2px solid var(--color-primary);

  ${({ isFull, isSubmit }) => {
    const shouldBeFull = isFull || isSubmit

    return css`
      background-color: ${shouldBeFull ? 'var(--color-primary)' : 'transparent'};
      color: ${shouldBeFull ? 'var(--color-white)' : 'var(--color-primary)'};

      :hover {
        background-color: ${shouldBeFull ? 'var(--color-dark-primary)' : 'var(--color-primary)'};
        color: ${!shouldBeFull && 'var(--color-white)'};
      }
    `
  }}
`

export default Button
