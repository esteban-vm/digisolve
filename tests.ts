import type { ReactElement } from 'react'
import { create, type ReactTestRendererJSON, type TestRendererOptions } from 'react-test-renderer'

const customCreate = (ui: ReactElement, options?: TestRendererOptions) => {
  return create(ui, options).toJSON() as ReactTestRendererJSON
}

export { customCreate as create }
export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
