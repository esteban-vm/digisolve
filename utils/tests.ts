import type { ReactElement } from 'react'
import { create } from 'react-test-renderer'

const customCreate = (element: ReactElement) => create(element).toJSON()
export { customCreate as create }

export * from '@testing-library/react'
export * from 'jest-axe'
export { default as userEvent } from '@testing-library/user-event'
