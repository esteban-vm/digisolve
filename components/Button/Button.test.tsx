import { render, cleanup, screen, create, axe } from '@/tests'
import Button from './Button'

describe('ðŸ§ª BUTTON:', () => {
  describe('render tests:', () => {
    let container: HTMLElement

    beforeEach(() => {
      void ({ container } = render(<Button text='test' />))
    })

    afterEach(cleanup)

    it('should be accessible', async () => {
      const results = await axe(container)
      expect(results).toHaveNoViolations()
    })

    it('the button should be rendered', () => {
      const button = screen.getByRole('button', { name: 'test' })
      expect(button).toBeInTheDocument()
      expect(button).toBeVisible()
    })
  })

  describe('style tests:', () => {
    let tree: ReturnType<typeof create>

    describe('with basic button:', () => {
      beforeEach(() => {
        tree = create(<Button text='test' />)
      })

      it('should render with correct styles', () => {
        expect(tree).toMatchSnapshot()
      })

      it('should have correct text and background color', () => {
        const target = ':hover'
        expect(tree).toHaveStyleRule('background-color', 'var(--color-primary)', { target })
        expect(tree).toHaveStyleRule('color', 'var(--color-white)', { target })
      })
    })

    describe('with full button:', () => {
      beforeEach(() => {
        tree = create(<Button text='test' isFull />)
      })

      it('should render with correct styles', () => {
        expect(tree).toMatchSnapshot()
      })

      it('should have correct background color', () => {
        expect(tree).toHaveStyleRule('background-color', 'var(--color-dark-primary)', { target: ':hover' })
      })
    })

    describe('with submit button', () => {
      it('should render with correct styles', () => {
        tree = create(<Button text='test' isSubmit />)
        expect(tree).toMatchSnapshot()
      })
    })
  })
})
