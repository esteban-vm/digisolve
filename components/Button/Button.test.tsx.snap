// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ§ª BUTTON: style tests: with basic button: should render with correct styles 1`] = `
.emotion-0 {
  cursor: pointer;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  padding: 10px 30px;
  border-radius: 200px;
  font-weight: 500;
  outline: none;
  -webkit-transition: background-color 500ms;
  transition: background-color 500ms;
  border: 2px solid var(--color-primary);
  background-color: transparent;
  color: var(--color-primary);
}

.emotion-0:hover {
  background-color: var(--color-primary);
  color: var(--color-white);
}

<button
  className="emotion-0"
  onClick={[Function]}
  type="button"
>
  test
</button>
`;

exports[`ðŸ§ª BUTTON: style tests: with full button: should render with correct styles 1`] = `
.emotion-0 {
  cursor: pointer;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  padding: 10px 30px;
  border-radius: 200px;
  font-weight: 500;
  outline: none;
  -webkit-transition: background-color 500ms;
  transition: background-color 500ms;
  border: 2px solid var(--color-primary);
  background-color: var(--color-primary);
  color: var(--color-white);
}

.emotion-0:hover {
  background-color: var(--color-dark-primary);
}

<button
  className="emotion-0"
  onClick={[Function]}
  type="button"
>
  test
</button>
`;

exports[`ðŸ§ª BUTTON: style tests: with submit button should render with correct styles 1`] = `
.emotion-0 {
  cursor: pointer;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  padding: 10px 30px;
  border-radius: 200px;
  font-weight: 500;
  outline: none;
  -webkit-transition: background-color 500ms;
  transition: background-color 500ms;
  border: 2px solid var(--color-primary);
  background-color: var(--color-primary);
  color: var(--color-white);
}

.emotion-0:hover {
  background-color: var(--color-dark-primary);
}

<button
  className="emotion-0"
  type="submit"
>
  test
</button>
`;
